add forecast api dao

adamkorynta opened the pull request.

adamkorynta committed:
add forecast api dao

rma-rripken commented on cwms-data-api/src/main/java/cwms/cda/ApiServlet.java:
```diff
@@ -285,6 +286,11 @@ public void init() {
                     logger.atInfo().withCause(e).log(re.toString(), e);
                     ctx.status(HttpServletResponse.SC_CONFLICT).json(re);
                 })
+                .exception(NoDataFoundException.class, (e, ctx) -> {
+                    CdaError re = new CdaError("Not Found.");
+                    logger.atInfo().withCause(e).log(re.toString());
+                    ctx.status(HttpServletResponse.SC_NOT_FOUND).json(re);
```
There is already a NotFoundException - rather not bring org.jooq into this

adamkorynta replied:
removed it

rma-rripken commented on cwms-data-api/src/main/java/cwms/cda/ApiServlet.java:
```diff
@@ -436,10 +442,10 @@ protected void configureRoutes() {
-//                new ForecastInstanceController(metrics), requiredRoles,5, TimeUnit.MINUTES);
-//        cdaCrudCache("/forecast-spec/{" + Controllers.NAME + "}",
-//                new ForecastSpecController(metrics), requiredRoles,5, TimeUnit.MINUTES);
+        cdaCrudCache("/forecast-instance/{" + Controllers.NAME + "}",
+                new ForecastInstanceController(metrics), requiredRoles,5, TimeUnit.MINUTES);
+        cdaCrudCache("/forecast-spec/{" + Controllers.NAME + "}",
+                new ForecastSpecController(metrics), requiredRoles,5, TimeUnit.MINUTES);
```
You'll need the @WebServlet part too

adamkorynta replied:
thanks!

adamkorynta committed:
revert to a single forecast location per spec

implement joins in dao for the forecast info and forecast time series extents

adamkorynta committed:
bug fixes from integration testing

adamkorynta committed:
error handling cleanup

rma-rripken commented on cwms-data-api/src/main/java/cwms/cda/data/dao/JooqDao.java:
```diff
@@ -173,6 +175,14 @@ public void accept(org.jooq.Context<?> ctx) {
 
+    protected static Condition filterExact(Field<String> field, String filter) {
+        if (filter == null) {
+            return DSL.trueCondition();
```
I think DSL.noCondition is what you want here

adamkorynta replied:
Oh didn't know that one! Updated

adamkorynta committed:
spec integration tests now passing

currently need to creat time series before storing the spec
also need to fully truncate the at_fcst_time_series table due to cyclic references when trying to cleanup data

adamkorynta committed:
add file download from url, use straight sql instead of jooq to avoid eager blob

adamkorynta committed:
update tests to handle null

rma-rripken commented on cwms-data-api/src/main/java/cwms/cda/data/dao/ForecastSpecDao.java:
```diff
@@ -1,38 +1,124 @@
+        AV_FCST_LOCATION loc = AV_FCST_LOCATION.AV_FCST_LOCATION;
+        //Group all the timeseries ids into a "\n" delimited list
+        Table<Record2<String, String>> tsidTable = dsl.select(timeSeries.FCST_SPEC_CODE,
+                        DSL.listAgg(timeSeries.CWMS_TS_ID, "\n")
```
interesting. haven't seen this done before

adamkorynta replied:
AI prompt gave me that suggestion. I saw it also has a usage in the cwms_util package

rma-rripken commented on cwms-data-api/src/test/java/cwms/cda/api/ForecastInstanceControllerTestIT.java:
```diff
@@ -150,55 +300,69 @@ void test_create_get_delete_get() throws IOException {
+            .body("notes", equalTo("test notes"))
+//                .body("metadata.key1", equalTo("value1"))
+//                .body("metadata.key2", equalTo("value2"))
+//                .body("metadata.key3", equalTo("value3"))
```
is metadata going away or its tested somewhere else?

adamkorynta replied:
It currently doesn't work in the schema: https://bitbucket.hecdev.net/projects/CWMS/repos/cwms_database/pull-requests/326/overview?commentId=17908

rma-rripken commented on cwms-data-api/src/main/java/cwms/cda/data/dao/ForecastInstanceDao.java:
```diff
@@ -1,37 +1,321 @@
+        //Will throw a NotFoundException if instance doesn't exist
+        ReplaceUtils.OperatorBuilder noopUrlBuilder = new ReplaceUtils.OperatorBuilder().withTemplate("")
+                .withOperatorKey("{noop}");
+        getForecastInstance(0, noopUrlBuilder, officeId, specId, designator, forecastDate, issueDate);
```
This is weird to me but I think I get it.  I would have put more thought into OperatorBuilder if I knew it was going to be part of our api's.

adamkorynta replied:
yeah, I was just so relieved to get my giant query going, that I didn't spend time on the replaceutils stuff

rma-rripken commented on cwms-data-api/src/main/java/cwms/cda/helpers/ReplaceUtils.java:
```diff
@@ -38,15 +38,19 @@ public static String replace(@NotNull String template, @NotNull String key, Stri
      * @throws NullPointerException if the template is null
      */
-    public static String replace(@NotNull String template, @NotNull String key, String value, boolean encode){
+    public static String replace(@NotNull String template, String key, String value, boolean encode){
```
What ends up calling this with key==null ?  Having this handle null does seem clean so I'm not opposed its just weird that something wants to do replacement and doesn't know the key its replacing

adamkorynta replied:
I use the same pattern for getAll and getOne. In the case of getAll we have a template since we don't know which row spec, but in getOne we know the row spec. I could template them both but that seemed confusing to me too.

adamkorynta committed:
bug fixes for spec getAll and optional params

adamkorynta committed:
updates found from client testing bug fixes
